src/shogun/labels/MappedLabels.h

```diff
@@ -0,0 +1,127 @@
+/*
+ * This software is distributed under BSD 3-clause license (see LICENSE file).
+ *
+ * Authors: Heiko Strathmann
+ */
+
+#ifndef MAPPED_LABELS_H
+#define MAPPED_LABELS_H
+
+#ifndef SWIG
+#include <shogun/lib/config.h>
+
+#include <shogun/lib/common.h>
+
+#include <shogun/labels/DenseLabels.h>
+#include <shogun/base/class_list.h>
+
+namespace shogun
+{
+
+typedef std::map<float64_t, float64_t> LabelMap;
+
+template <typename T,
+	      std::enable_if_t<
+	          std::is_base_of<
+	              CDenseLabels, typename std::remove_pointer<T>::type>::value,
+	          T>* = nullptr>
+class MappedLabels : public T
+{
+	template<typename U, typename S> friend class MappedLabelsFactory;
+public:
+	MappedLabels() : T() {}// for class_list.h :(
+	MappedLabels(CLabels* orig) : T()
+	{
+		auto dense = orig->as<CDenseLabels>();
+		ASSERT(dense);
+		m_orig = dense;
+		SG_REF(m_orig);
+	}
+
+	virtual ~MappedLabels()
+	{
+		SG_UNREF(m_orig);
+	}
+
+	// access now needs to be virtual in CDenseLabels
+	virtual SGVector<float64_t> get_labels() const
+	{
+		auto orig = m_orig->get_labels();
+		if (!m_to_internal.size())
+			return orig;
+
+		SGVector<float64_t> converted(orig.size());
+		auto lab = m_orig->get_labels(); // TODO use iterator on label directly
+		std::transform(lab.begin(), lab.end(), converted.begin(),
+				[this](float64_t l) {
+					return this->m_to_internal.at(l);
+				});
+		return converted;
+	}
+
+	// access now needs to be virtual in CDenseLabels
+	virtual float64_t get_label(index_t i) const
+	{
+		auto orig = m_orig->get_label(i);
+		return m_to_internal.size() ? m_to_internal.at(orig) : orig;
+	}
+
+	virtual int32_t get_num_labels() const { return m_orig->get_num_labels(); }
+
+	CDenseLabels* invert(const CLabels* labels) const
+	{
+		auto dense = labels->as<CDenseLabels>();
+		ASSERT(dense); // internal error
+
+		auto orig = dense->get_labels();
+		auto inverted = SGVector<float64_t>(orig.size());
+		std::transform(orig.begin(), orig.end(), inverted.begin(),
+				[this](float64_t l) {
+					return this->m_from_internal.at(l);
+				});
+
+		auto result = create_object<CDenseLabels>(m_orig->get_name());
+		SG_REF(result);
+		result->set_labels(inverted);
+		return result;
+	}
+
+	// trait to only define for mixin instantiations
+	template<class MAPPED_CLASS, std::enable_if_t<std::is_base_of<T, MAPPED_CLASS>::value, MAPPED_CLASS>* = nullptr>
+	static T* wrap_if_necessary(CLabels* orig)
+	{
+		auto casted = dynamic_cast<T*>(orig);
+		if (casted)
```

Comment:
you can move the line above into the `if` so it stays in its own scope, i.e. `if (auto* casted = dynamic_cast<T*>(orig)) return casted`